text/0000-socket-timeouts.md

```diff
@@ -0,0 +1,142 @@
+- Feature Name: socket_timeouts
+- Start Date: 2015-04-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add sockopt-style timeouts to `std::net` types.
+
+# Motivation
+
+Currently, operations on various socket types in `std::net` block
+indefinitely (i.e., until the connection is closed or data is
+transferred). But there are many contexts in which timing out a
+blocking call is important.
+
+The [goal of the current IO system][io-reform] is to gradually expose
+cross-platform, blocking APIs for IO, especially APIs that directly
+correspond to the underlying system APIs. Sockets are widely available
+with nearly identical system APIs across the platforms Rust targets,
+and this includes support for timeouts via [sockopts][sockopt].
+
+So timeouts are well-motivated and well-suited to `std::net`.
+
+# Detailed design
+
+The proposal is to *directly expose* the timeout functionality
+provided by [`setsockopt`][sockopt], in much the same way we currently
+expose functionality like `set_nodelay`:
+
+```rust
+impl TcpStream {
+    pub fn set_read_timeout(&self, dur: Option<Duration>) -> io::Result<()> { ... }
+    pub fn read_timeout(&self) -> Option<Duration>;
```

Comment:
I personally see `Result<Option<Duration>>` as quite bad (too much nesting). I re-consulted the documentation for `getsockopt` and [it states](http://pubs.opengroup.org/onlinepubs/009695399/functions/getsockopt.html)

> Note that not all implementations allow this option to be retrieved.

This means that a successful `set_write_timeout` may not be followed by a successful `write_timeout` so you're probably not just `try!`-ing away the error in the robust case anyway.